Add necessary native-image configuration files

dreamlike-ocean opened the pull request:
Add necessary native-image configuration files for AdaptivePoolingAllocator,Epoll,Nio,IO_Uring,Quic

Motivation:
IoUring and Adaptive allocator currently doesn't work out of the box on native-image without passing in additional flags and metadata.

Modification:

Add necessary native-image configuration files for AdaptivePoolingAllocator,Epoll,Nio,IO_Uring,Quic

Result:

Fixes #14903

dreamlike-ocean commented:
> @dreamlike-ocean As this configurations are new should we follow the latest simplified reachability metadata where only one file is needed - reachability-metadata.json?
> https://www.graalvm.org/latest/reference-manual/native-image/metadata/#specifying-metadata-with-json

The new version of GraalVM supports single configuration files. Perhaps i can provide both types of files simultaneously?

normanmaurer commented:
@violetagg @dreamlike-ocean what is the status of this one ?

dreamlike-ocean commented:
> @violetagg @dreamlike-ocean what is the status of this one ?

I currently support the native image meta file format of the old version.
If necessary to switch to the new version format, we might need to regenerate the entire Netty project—or perhaps we can do it later.  
Both formats are supported in the newer GraalVM versions.

dreamlike-ocean commented:
@violetagg could you review it?

violetagg commented:
I’ll review it once I return from vacation

On Sun, 30 Mar 2025 at 14:02, Mengyang Li ***@***.***> wrote:

> @violetagg <https://github.com/violetagg> could you review it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/netty/netty/pull/14928#issuecomment-2764553818>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFKCVJVFECE4NQ6ONTNQ3T2W7TOTAVCNFSM6AAAAABZCS3X6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUGU2TGOBRHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: dreamlike-ocean]*dreamlike-ocean* left a comment
> (netty/netty#14928)
> <https://github.com/netty/netty/pull/14928#issuecomment-2764553818>
>
> @violetagg <https://github.com/violetagg> could you review it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/netty/netty/pull/14928#issuecomment-2764553818>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFKCVJVFECE4NQ6ONTNQ3T2W7TOTAVCNFSM6AAAAABZCS3X6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUGU2TGOBRHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

chrisvest commented:
@dreamlike-ocean please rebase so you get https://github.com/netty/netty/pull/14987, and then add some usage to `testsuite-native-image` (or an appropriate sibling module) to get some coverage and confidence that it now works.

violetagg commented:
> > @violetagg @dreamlike-ocean what is the status of this one ?
> 
> I currently support the native image meta file format of the old version. If necessary to switch to the new version format, we might need to regenerate the entire Netty project—or perhaps we can do it later. Both formats are supported in the newer GraalVM versions.

Ok, let's use the old format

yawkat commented:
fwiw, i asked the native image team and they say there are no plans to drop the old format any time soon, so we can continue using it.

dreamlike-ocean commented:
Thanks for waiting, and sorry for the delay. Appreciate your effort in reviewing it again!
@violetagg @normanmaurer @chrisvest
1. Add HttpNativeServer for testing various transport native implementations  
2. Add AdaptivePoolingAllocatorSubstitution to support native image compatibility for AdaptivePoolingAllocator

normanmaurer commented:
@dreamlike-ocean seems like the Graal build failed... PTAL

dreamlike-ocean commented:
> @dreamlike-ocean seems like the Graal build failed... PTAL

I’m checking the build output,
and it looks like this is what’s causing the issue:
``` txt
2025-04-15T17:24:50.4980660Z [ERROR] Failed to execute goal on project netty-testsuite-native-image: Could not resolve dependencies for project io.netty:netty-testsuite-native-image:jar:4.2.1.Final-SNAPSHOT
2025-04-15T17:24:50.4982291Z [ERROR] dependency: io.netty:netty-transport-native-epoll:jar:linux-aarch_64:4.2.1.Final-SNAPSHOT (runtime)
2025-04-15T17:24:50.4987223Z [ERROR]   Could not find artifact io.netty:netty-transport-native-epoll:jar:linux-aarch_64:4.2.1.Final-SNAPSHOT in sonatype-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots)
2025-04-15T17:24:50.4989236Z [ERROR] dependency: io.netty:netty-transport-native-io_uring:jar:linux-aarch_64:4.2.1.Final-SNAPSHOT (runtime)
2025-04-15T17:24:50.4990991Z [ERROR]   Could not find artifact io.netty:netty-transport-native-io_uring:jar:linux-aarch_64:4.2.1.Final-SNAPSHOT in sonatype-nexus-snapshots (https://oss.sonatype.org/content/repositories/snapshots)
2025-04-15T17:24:50.4992358Z [ERROR] -> [Help 1]
```

Let me find a way to fix it — I’m working on it.

dreamlike-ocean commented:
Finally! all checks have passed!
Looks like the native image testsuite is running well.
I’ve completed the following tasks:
	1.	Fixed the issue where the testsuite native image could not produce multiple native image artifacts.
	2.	Added missing reflection configurations for NIO and Epoll.
	3.	Introduced native image configuration for io_uring and adaptiveAllocator.
	4.	Added missing native image configuration for codec-quic.

chrisvest commented on testsuite-native-image/pom.xml:
```diff
@@ -61,7 +63,21 @@
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-codec-classes-quic</artifactId>
+      <version>${netty.version}</version>
```
We can just inline `${project.version}` in these three.

chrisvest commented:
@dreamlike-ocean Thanks!
